File: Python/render.py
"""渲染函数模块

这个文件是你参赛时使用的、允许修改的文件。
分别在 `rd` `gr` `bl` 里填写你的渲染函数。
渲染时，会在同一像素位置 (x, y) 上分别调用
这三个函数获取像素位置的颜色 (R, G, B)。
"""

from math import *

from handy_math import *


FPS = 25
FPS_INV = 1.0 / FPS
TMAX = 6.0
DIM = 512
DM1 = (DIM - 1)

# image size
N = DIM
N1 = DIM - 1


def rd(x: int, y: int, t: float) -> int:
    """像素红色

    参数:
        x (int): x 坐标
        y (int): y 坐标
        t (float): 时间 (单位：秒)

    返回值:
        int: 红色值 (0-255)
    """
    # 在这里填写你的代码
    return 0


def gr(x: int, y: int, t: float) -> int:
    """像素绿色

    参数:
        x (int): x 坐标
        y (int): y 坐标
        t (float): 时间 (单位：秒)

    返回值:
        int: 绿色值 (0-255)
    """
    # 在这里填写你的代码
    return 0


def bl(x: int, y: int, t: float) -> int:
    """像素蓝色

    参数:
        x (int): x 坐标
        y (int): y 坐标
        t (float): 时间 (单位：秒)

    返回值:
        int: 蓝色值 (0-255)
    """
    # 在这里填写你的代码
    return 0

File: JavaScript/scripts/render.ts
import '../utils/global';
import * as jsR from '../src/js';
import * as tsR from '../src/ts';

import { cpus } from 'os';
import Piscina from 'piscina';
import { program } from 'commander';
import GIFEncoder from 'gifencoder';
import { createWriteStream } from 'fs';
import { dirname, resolve } from 'path';
import { SingleBar } from 'cli-progress';
import { pipeline } from 'stream/promises';

program
    .option('-t, --threads <number>', 'number of threads', cpus().length.toString())
    .option('-f, --from <number>', 'start frame', '0')
    .option('-l, --length <number>', 'number of frames', (TMAX * FPS).toString())
    .argument('<output>', 'Output file')
    .action(async (output) => {
        console.log('BMAC JS/TS Build Script');
        let rename: 'ts' | 'js';
        const threads = Number(program.opts().threads) || cpus().length * 2;
        const startFrame = Number(program.opts().from) || 0;
        const totalFrames = Number(program.opts().length) || TMAX * FPS;
        if (tsR.useThis) {
            rename = 'ts';
        } else if (jsR.useThis) {
            rename = 'js';
        } else {
            console.log('Error: No renderer selected');
            process.exit(2);
        }
        // convert import.meta.url to __dirname
        let __dirname = dirname(decodeURI(new URL(import.meta.url).pathname));
        // if windows
        if (process.platform === 'win32') {
            __dirname = __dirname.substr(1);
        }
        const p = new Piscina({
            filename: resolve(__dirname, './worker.js'),
            maxThreads: threads,
        });
        console.log(
            `Rendering frmaes ${startFrame}-${
                startFrame + totalFrames - 1
            } with ${threads} threads with ${rename.toUpperCase()} renderer...`,
        );
        const encoder = new GIFEncoder(DIM, DIM);
        const stream = encoder.createReadStream();
        encoder.start();
        encoder.setRepeat(0); // 0 for repeat, -1 for no-repeat
        encoder.setDelay(FPS_INV * 1000); // frame delay in ms
        encoder.setQuality(10);
        const bar = new SingleBar(
            {},
            {
                format: '[{bar}] {percentage}% | ETA: {eta}s | {value}/{total}',
                barCompleteChar: '=',
                barIncompleteChar: '-',
            },
        );
        bar.start(totalFrames, 0);
        const promises = [];
        let nowtime = Date.now();
        let renderedFrame = startFrame;
        const cachedFrames = new Map<number, Uint8ClampedArray>();
        for (let fn = startFrame; fn < totalFrames; fn++) {
            promises.push(
                p.run({ rename, fn }).then((imgBuf) => {
                    if (renderedFrame == fn) {
                        encoder.addFrame(imgBuf);
                        bar.increment();
                        renderedFrame++;
                        while (cachedFrames.has(renderedFrame)) {
                            encoder.addFrame(cachedFrames.get(renderedFrame) as any);
                            cachedFrames.delete(renderedFrame);
                            renderedFrame++;
                            bar.increment();
                        }
                    } else if (renderedFrame < fn) {
                        cachedFrames.set(fn, imgBuf);
                    } else {
                        throw new Error('Invalid frame sequence');
                    }
                }),
            );
        }
        await Promise.all(promises);
        encoder.finish();
        bar.update(totalFrames);
        bar.stop();
        console.log('Saving GIF...');
        await pipeline(stream, createWriteStream(output));
        console.log('Done in ' + ((Date.now() - nowtime) / 1000).toFixed(2) + 's.');
    })
    .parse(process.argv);
